secret_store/src/key_server_cluster/cluster_message_processor.rs

```diff
@@ -0,0 +1,357 @@
+		&self,
+		sessions: &ClusterSessionsContainer<S, SC, D>,
+		connection: Arc<Connection>,
+		mut message: Message,
+	) -> Option<Arc<S>>
+		where
+			Message: IntoSessionId<S::Id>
+	{
+		// get or create new session, if required
+		let mut sender = connection.node_id().clone();
+		let session = self.prepare_session(sessions, &sender, &message);
+		// send error if session is not found, or failed to create
+		let session = match session {
+			Ok(session) => session,
+			Err(error) => {
+				// this is new session => it is not yet in container
+				warn!(target: "secretstore_net",
+					"{}: {} session read error '{}' when requested for session from node {}",
+					self.self_key_pair.public(), S::type_name(), error, sender);
+				if !message.is_error_message() {
+					let qed = "session_id only fails for cluster messages;
+						only session messages are passed to process_message;
+						qed";
+					let session_id = message.into_session_id().expect(qed);
+					let session_nonce = message.session_nonce().expect(qed);
+
+					connection.send_message(SC::make_error_message(session_id, session_nonce, error));
+				}
+				return None;
+			},
+		};
+
+		let session_id = session.id();
+		let mut is_queued_message = false;
+		loop {
+			let message_result = session.on_message(&sender, &message);
+			match message_result {
+				Ok(_) => {
+					// if session is completed => stop
+					if session.is_finished() {
+						info!(target: "secretstore_net",
+							"{}: {} session completed", self.self_key_pair.public(), S::type_name());
+						sessions.remove(&session_id);
+						return Some(session);
+					}
+
+					// try to dequeue message
+					match sessions.dequeue_message(&session_id) {
+						Some((msg_sender, msg)) => {
+							is_queued_message = true;
+							sender = msg_sender;
+							message = msg;
+						},
+						None => return Some(session),
+					}
+				},
+				Err(Error::TooEarlyForRequest) => {
+					sessions.enqueue_message(&session_id, sender, message, is_queued_message);
+					return Some(session);
+				},
+				Err(err) => {
+					warn!(
+						target: "secretstore_net",
+						"{}: {} session error '{}' when processing message {} from node {}",
+						self.self_key_pair.public(),
+						S::type_name(),
+						err,
+						message,
+						sender);
+					session.on_session_error(self.self_key_pair.public(), err);
+					sessions.remove(&session_id);
+					return Some(session);
+				},
+			}
+		}
+	}
+
+	/// Get or insert new session.
+	fn prepare_session<S: ClusterSession, SC: ClusterSessionCreator<S, D>, D>(
+		&self,
+		sessions: &ClusterSessionsContainer<S, SC, D>,
+		sender: &NodeId,
+		message: &Message
+	) -> Result<Arc<S>, Error>
+		where
+			Message: IntoSessionId<S::Id>
+	{
+		fn requires_all_connections(message: &Message) -> bool {
+			match *message {
+				Message::Generation(_) => true,
+				Message::ShareAdd(_) => true,
+				Message::ServersSetChange(_) => true,
```

Comment:
Could be a oneliner `	Message::Generation(_) | Message::ShareAdd(_) | Message::ServersSetChange(_) => true`